### ISSUE TITLE
NodeJS

### ISSUE THREAD

**Issue by tonyunreal** [ThatGuySam/doesitarm#299]
**The official name of the app**
Node.js

**Is there a supported build available on a public/release channel?**
No

**Proposed App Status**
Yes, full native support as of v15, requires building from source

**Proposed App Category**
Developer Tools

**Related Issue Tracker Link or discussion**
Official discussion: https://github.com/nodejs/TSC/issues/886
User report on nvm build working: https://blockforums.org/topic/453-nodejs-on-apple-silicon-m1

**Comment by tonyunreal** [comment]
Never mind, I just realized it was already reported in readme.md . Maybe this issue could serve further discussions if needed.

**Comment by tinymachine** [comment]
@ThatGuySam, the latest LTS of Node (14.15.1) works fine for me on an M1 MacBook Air via (I presume) Rosetta 2, so perhaps the status should be changed from `⏹` to `✳️ Yes, works via Rosetta 2`?

**Comment by ThatGuySam** [comment]
@tinymachine You can check if it's running in with Rosetta(Intel) or Native(Apple) by checking Activity Monitor. 

Feel free to drop a screenshot in here if you do.

**Comment by tinymachine** [comment]
Ah, of course -- thanks for the tip! 14.15.1 is currently working fine via Rosetta 2:
<img width="898" alt="Screen Shot 2020-11-23 at 9 49 04 PM" src="https://user-images.githubusercontent.com/171986/100053837-b7363880-2dd5-11eb-9374-baaaa021adf0.png">

**Comment by jonathantneal** [comment]
To whom it serves, you can install Node v15.3.0 on Apple Silicon.

<img width="953" alt="image" src="https://user-images.githubusercontent.com/188426/100143608-93213880-2e63-11eb-9285-3a80d9bbb915.png">

here is how I did it:

```sh
# from the temporary directory, download and untar Node v15.3.0
echo "Downloading Node v15.3.0 (1/4)..."
cd $TMPDIR
curl -#OL https://nodejs.org/dist/v15.3.0/node-v15.3.0.tar.gz
tar xf node-v15.3.0.tar.gz

# configure the Node installation for arm64 on macOS
echo "Installing Node v15.3.0 (2/4)..."
cd $TMPDIR/node-v15.3.0
./configure --dest-cpu=arm64 --dest-os=mac

# ensure node is built with m1 xcode tools and perform the installation (this takes about 10 minutes)
BACKUP_PATH=$PATH PATH=/usr/bin:$PATH
sudo make install -j8 V= DESTCPU="arm64" ARCH="arm64" VARIATION="" DISTYPE="release" RELEASE_URLBASE="https://nodejs.org/download/release/"
PATH=BACKUP_PATH

# ensure npm has global installation permissions
echo "Permitting Node v15.3.0 (3/4)..."
sudo chown -R $(whoami) $(npm config get prefix)/{lib/node_modules,bin,share}

# update npm to verify the installation and permissions
echo "Updating NPM (4/4)..."
npm install -g npm
```

**Comment by ThatGuySam** [comment]
@tinymachine You have a good point about changing the status to "✳️ Yes, works via Rosetta 2". 

However, v15 can run natively on ARM. 

So for clarity let's do this: 
The Status is based on the current Active LTS version. 

So until Node 16 is officially released the status will be whatever v14 is. 

If v14 gets native arm support released then we'll update the status to Fully Native.

**Comment by ThatGuySam** [comment]
Here's the current support as of Nov 24, 2020

v14: ✳️ Yes, works via Rosetta 2
v15: ✅ Native support with a workaround

**Comment by ThatGuySam** [comment]
I'll link directly to that ^^ comment so people can get finer details on version support

**Comment by ThatGuySam** [comment]
Update is live!

https://doesitarm.com/app/nodejs/

**Comment by Lee-Jp** [comment]
What does Node v6 support look like?

**Comment by elliottkember** [comment]
@jonathantneal Thank you for your snippet! I tried it, but something went wrong and I get this:

```
  CXX(host) /private/var/folders/vx/6cq64y3d30dgs9fglm2456rr0000gn/T/node-v15.3.0/out/Release/obj.host/v8_base_without_compiler/gen/torque-output-root/torque-generated/class-verifiers-tq.o
  CXX(host) /private/var/folders/vx/6cq64y3d30dgs9fglm2456rr0000gn/T/node-v15.3.0/out/Release/obj.host/v8_base_without_compiler/gen/torque-output-root/torque-generated/factory-tq.o
  CXX(host) /private/var/folders/vx/6cq64y3d30dgs9fglm2456rr0000gn/T/node-v15.3.0/out/Release/obj.host/v8_base_without_compiler/gen/torque-output-root/torque-generated/objects-printer-tq.o
  CXX(host) /private/var/folders/vx/6cq64y3d30dgs9fglm2456rr0000gn/T/node-v15.3.0/out/Release/obj.host/v8_base_without_compiler/deps/v8/src/heap/base/stack.o
  CXX(host) /private/var/folders/vx/6cq64y3d30dgs9fglm2456rr0000gn/T/node-v15.3.0/out/Release/obj.host/v8_base_without_compiler/deps/v8/src/heap/base/asm/x64/push_registers_asm.o
<inline asm>:4:8: error: unknown token in expression
  push %rbp
       ^
<inline asm>:4:8: error: invalid operand
  push %rbp
       ^
<inline asm>:5:7: error: unknown token in expression
```

followed by a bunch more similar errors.

**Comment by AdamGerthel** [comment]
Node v15, which is the "Current" version, works without Rosetta:

![Screenshot 2020-12-28 at 19 59 24](https://user-images.githubusercontent.com/1490370/103237213-3247bd00-4947-11eb-82ac-ae3c8110eae1.png)

**Comment by richiksc** [comment]
I think this should be marked as Full Native Apple Silicon support, or it should be split up into two entries. v15 is still a stable, full-featured release, and it's much like Ubuntu - 20.10 is still a stable release, it's just that 20.04 is LTS.

**Comment by ThatGuySam** [comment]
How about this: 
✅ Yes, Full Native Apple Silicon Support as of v15

**Comment by rsboarder** [comment]
@jonathantneal I tried to install using your snippet by it failed with this error

```
make[1]: /bin/sh: Bad address
make[1]: *** [/private/var/folders/d1/fhw7tc315yn66plw3zgg5d3c0000gn/T/node-v15.6.0/out/Release/obj.target/v8_initializers/gen/torque-output-root/torque-generated/../../deps/v8/src/builtins/ic-tq-csa.o] Error 1
make[1]: *** Waiting for unfinished jobs....
rm a25728ab8eb3da324c06ae888715977bd383b7a9.intermediate 7961cfc3c172c83354f810b2ae2142c26e5ad587.intermediate cc4cc11b0d60296a583be691e4c222369fab8c07.intermediate
make: *** [node] Error 2
```

Maybe you know why is it so?

**Comment by AdamGerthel** [comment]
> @jonathantneal I tried to install using your snippet by it failed with this error
> 
> ```
> make[1]: /bin/sh: Bad address
> make[1]: *** [/private/var/folders/d1/fhw7tc315yn66plw3zgg5d3c0000gn/T/node-v15.6.0/out/Release/obj.target/v8_initializers/gen/torque-output-root/torque-generated/../../deps/v8/src/builtins/ic-tq-csa.o] Error 1
> make[1]: *** Waiting for unfinished jobs....
> rm a25728ab8eb3da324c06ae888715977bd383b7a9.intermediate 7961cfc3c172c83354f810b2ae2142c26e5ad587.intermediate cc4cc11b0d60296a583be691e4c222369fab8c07.intermediate
> make: *** [node] Error 2
> ```
> 
> Maybe you know why is it so?

You don't need a special snippet, you can just install it the regular way.

**Comment by rsboarder** [comment]
@AdamGerthel you installing via brew?

**Comment by toddjonker** [comment]
I'd really appreciate more explicit guidance on older releases; some of us are stuck on old things for various reasons and it would be nice to know whether that's going to block M1 adoption.

**Comment by ThatGuySam** [comment]
@toddjonker What do you think would be a good solution?

**Comment by AdamGerthel** [comment]
> @AdamGerthel you installing via brew?

Nope, I installed via NVM. A month ago only v15 of node was supported but I installed v12 a few days ago without any problems.

Some libraries might not work properly, for instance I've had trouble with node-sass. That's not something that a list like this can keep track of imo. 

It's definitely possible to simply install a different version of node in a Rosetta-launched terminal if needed.

**Comment by moffpage** [comment]
Why is it still an issue?

**Comment by ThatGuySam** [comment]
@moffpage This is just an older discussion from before the Node 16 release. 

I'll close it now.
